### PR TITLE
OpInfo tests for `(svd|pca)_lowrank`

### PR DESCRIPTION
**Pull request by nikitaved**
As per title.

While working on this I have discovered several issues with these methods related to grad instabilities. I will file them and link here later. These were quite painful to force to pass all the tests with these discovered issues, sorry for the delay, @mruberry!

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -7589,12 +7680,13 @@ def reference_mse_loss(input, target, reduction="mean"):
...
 
-def wrapper_set_seed(op, input, *args, **kwargs):
+def wrapper_set_seed(op, *args, **kwargs):
     """Wrapper to set seed manually for some functions like dropout
     See: https://github.com/pytorch/pytorch/pull/62315#issuecomment-896143189 for more details.
     """
-    torch.manual_seed(42)
-    return op(input, *args, **kwargs)
+    with freeze_rng_state():
+        torch.manual_seed(42)
```
Comment (nikitaved) [inline torch/testing/_internal/common_methods_invocations.py:7688]: NOTE: `freeze_rng_state` is causing `test_composite_compliance` to fail.

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -2492,6 +2492,97 @@ def generate_samples():
 
     return list(generate_samples())
 
+
+def sample_inputs_singular_matrix_factors(op_info, device, dtype, requires_grad=False, **kwargs):
+
```
Comment (nikitaved) [inline torch/testing/_internal/common_methods_invocations.py:2497]: Add a comment about this function.

**Comment by nikitaved** [comment]
The newly introduced `test_composite_compliance` is failed thought the test base, hence the failures it seems.

File: test/test_fx_experimental.py [hunk 3]
```diff
@@ -1535,6 +1535,8 @@ def test_normalize_operator_exhaustive(self, device, dtype, op):
             "atleast_1d",
             "atleast_2d",
             "atleast_3d",
+            "svd_lowrank",
+            "pca_lowrank",
```
Comment (nikitaved) [inline test/test_fx_experimental.py:1539]: Add a comment explaining that OpInfo tests are using lambdas.

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -7754,12 +7845,21 @@ def reference_mse_loss(input, target, reduction="mean"):
...
+    # from torch.testing._internal.common_utils import freeze_rng_state
+    # with freeze_rng_state():
+    #     torch.manual_seed(42)
+    #     return op(*args, **kwargs)
+    # This code could be future-proof against accidental
+    # changes in the global random state.
+
     torch.manual_seed(42)
-    return op(input, *args, **kwargs)
+    return op(*args, **kwargs)
```
Comment (Lezcano) [inline torch/testing/_internal/common_methods_invocations.py:7862]: @mruberry You'll bump into this when reviewing this, but I'm interested. What are the implications of this?

Comment (nikitaved) [reply torch/testing/_internal/common_methods_invocations.py:7862]: I do not expect any issues as of now, given that seed is reset prior to each test, and all the samples are generated before calls to ops. This, however, might changed if and when OpInfo test will deal with generators and the processing pipeline changes to `sample1 -> op(sample1) -> sample2 -> op(sample2) -> ...`.

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -7886,12 +7986,21 @@ def reference_mse_loss(input, target, reduction="mean"):
         return se
 
 
-def wrapper_set_seed(op, input, *args, **kwargs):
+def wrapper_set_seed(op, *args, **kwargs):
     """Wrapper to set seed manually for some functions like dropout
     See: https://github.com/pytorch/pytorch/pull/62315#issuecomment-896143189 for more details.
     """
+    # freeze_rng_state causes test_composite_compliance to fail.
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:7993]: Nice comment

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -12014,6 +12123,38 @@ def ref_pairwise_distance(input1, input2):
...
+           ),
+           dtypes=floating_types(),
+           supports_out=False,
+           check_batched_grad=False,
+           check_batched_gradgrad=False,
+           supports_forward_ad=False,
+           sample_inputs_func=sample_inputs_svd_lowrank,
+           decorators=[skipCUDAIfNoMagmaAndNoCusolver, skipCUDAIfRocm, skipCPUIfNoLapack],
+           skips=(
+               # test does not work with passing lambda for op
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:12139]: Nice skips + comments

File: torch/testing/_internal/common_methods_invocations.py [hunk 7]
```diff
@@ -2593,6 +2593,106 @@ def generate_samples():
...
+            # TODO: fix bug in the documentation for svd_lowrank:
+            # M has to be (*, m, n), and not (*, 1, n) as written
+            # in the documentation
+            op_kwargs['M'] = make_tensor((*batch, m, n), device, dtype, requires_grad=requires_grad)
+            yield clone_sample(sample, **op_kwargs)
+
+    return list(generator())
+
+
+def chunk_iter(iterable, size):
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:2670]: Nit: we probably don't want this public at the file level

**Review by mruberry — APPROVED** [review]
Cool! Thanks @nikitaved!
